### DISCUSSION TITLE
possible bug with `removeNodes()` or its `v-model`

### DISCUSSION THREAD

**Discussion by ghiscoding** [bcakmakoglu/vue-flow#453] [Q&A] [answered]
Hello, I think I discovered a bug with `removeNodes()` which is that it doesn't seem to delete the last node when there's only 1 node left in the canvas. I also replicated the same bug in the [Save & Restore - Example](https://vueflow.dev/examples/save.html) by using the following code from the Example.

So in our use case, we use GUID as Id and when I delete the last Node and then add a new Node, it adds 2 Nodes to the canvas instead of 1 as it's supposed to do because the last node in still somehow in memory

##### App.vue
```ts
<script lang="ts" setup>
import { VueFlow } from '@vue-flow/core'
import { ref, watch } from 'vue'
import SaveRestoreControls from './Controls.vue'

const elements = ref([{ id: '1', label: 'Node 1', position: { x: 100, y: 100 } }])

// NEW: adding a value count & display it below
const count = ref(1)
watch(elements, (newValue) => {
  count.value = newValue.length;
})
</script>

<template>
  Elements count: {{count}}
  <VueFlow v-model="elements">
    <SaveRestoreControls />
  </VueFlow>
</template>
```
##### `Controls.vue`

```ts
<script setup>
import { Panel, PanelPosition } from '@vue-flow/additional-components'
import { useVueFlow } from '@vue-flow/core'

const flowKey = 'example-flow'

const { nodes, addNodes, setNodes, setEdges, dimensions, setTransform, toObject,getNodes, removeNodes } = useVueFlow()

const onSave = () => {
  localStorage.setItem(flowKey, JSON.stringify(toObject()))
}

const onRestore = () => {
  const flow = JSON.parse(localStorage.getItem(flowKey))

  if (flow) {
    const [x = 0, y = 0] = flow.position
    setNodes(flow.nodes)
    setEdges(flow.edges)
    setTransform({ x, y, zoom: flow.zoom || 0 })
  }
}

const onAdd = () => {
  const id = nodes.value.length + 1

  const newNode = {
    id: `${id}`,
    label: `Node ${id}`,
    position: { x: Math.random() * dimensions.value.width, y: Math.random() * dimensions.value.height },
  }

  addNodes([newNode])
}

function onDeleteLast() {
  removeNodes([`${getNodes.value.length}`]);
}
</script>

<template>
  <Panel :position="PanelPosition.TopRight" class="save-restore-controls">
    <button style="background-color: #33a6b8" @click="onSave">save</button>
    <button style="background-color: #113285" @click="onRestore">restore</button>
    <button style="background-color: #6f3381" @click="onAdd">add node</button>
    <button style="background-color: #6f3381" @click="onDeleteLast">delete last node</button>
  </Panel>
</template>
```
see animated gif

![firefox_SgVc8hEhUC](https://user-images.githubusercontent.com/643976/203427381-38ec044d-dfc2-4eb8-868e-33e1e8e61947.gif)

**Comment by bcakmakoglu** [comment]
Did you check if the element is actually still in the state? (like logging `getNodes.value`) 
or is your watcher just not triggering properly?^^

Try watching it deeply, i.e. `watch(value, () => {}, { deep: true })` and see if that triggers it properly.

**Reply by ghiscoding** [reply]
doesn't change anything and the watch is not really the problem, I added the watch just to see the element count (length) and so I don't need the watch it's just there to show you the problem. You can try it on your side with the code change I provided (make sure to also include proper `imports`), you should see the problem 

![firefox_MXvpDa0sVl](https://user-images.githubusercontent.com/643976/203428655-c1859750-9dc9-42f5-953f-626b7e825e0e.gif)

**Reply by ghiscoding** [reply]
I've edited the first post and added the entire code to test, it should be helpful for you. Thanks. 

NOTE: you won't see the problem that I'm having when you delete all Nodes and then click on "Add Node" to see 2 Nodes appear. You won't see this problem because "Add Node" uses the same Id: 1 when it starts, but I see the problem on my side because they are random guid.

**Reply by bcakmakoglu** [reply]
There's no bug with `removeNodes` as far as I can tell.
There is a bug with updating the v-model though, which is not really related to the `removeNodes` action but the model-value watcher.

That's why I asked if you can confirm that the element is or *is not* in the state :)

**Reply by ghiscoding** [reply]
> That's why I asked if you can confirm that the element is or is not in the state :)

oh ok that is probably the problem I'm facing, how do you propose to confirm this then?

this is currently what I have

```ts
<VueFlow v-model="elements" @edge-update="onEdgeUpdate" @connect="onConnect" :default-edge-options="{ type: 'custom' }" :experimentalFeatures="true">
```

**Reply by bcakmakoglu** [reply]
Don't worry, it *is* an issue that I'll have to fix just wanted to make sure I am looking at the same thing haha
Give me like 2 minutes to investigate, shouldn't be too hard to figure out 😄

**Reply by bcakmakoglu** [reply]
Quick note:

`:experimentalFeatures="true"` I don't think that's working btw

You should be accessing the experimental features flag via `useVueFlow` *or* by setting it with the `@pane-ready` event (which will give you a VueFlowStore instance as it's first argument).

So either this

```ts
useVueFlow({
  __experimentalFeatures: {
    nestedFlow: true,
  },
})
```

or 

```ts
const onPaneReady = (instance) => {
	instance.__experimentalFeatures.nestedFlow = true
}
```

**Reply by bcakmakoglu** [reply]
Also here's the fix for the issue with v-model: https://github.com/bcakmakoglu/vue-flow/pull/454

I'll release it tomorrow, if that is okay with you 😄 

you can try this docs link to see the fix in action: https://vue-flow-docs-git-fix-update-model-value-bcakmakoglu.vercel.app/
The docs always use the PR build so the fix is already available in the examples for that link

**Reply by ghiscoding** [reply]
sure take your time for a release, next day or week is fine, we're still in POC mode 😉 

thanks for the note about experimental, it seems that only the 2nd option works with TS, the 1st option throws a TS type error. I'm not sure if you exposed it in your interface? I'm on latest `v1.5.0`.. and that experimental flag doesn't seem to help so far, but that type of features will be implemented later so that can wait too

![image](https://user-images.githubusercontent.com/643976/203432189-3169abf7-205a-4be8-a56a-73b6c06b7f73.png)

**Reply by bcakmakoglu** [reply]
It's because you're passing an incomplete object.
There is no "options".
The options are simply an object that you pass as the first argument :)

```ts
// no `options: {}`
useVueFlow({
  __experimentalFeatures: {
    nestedFlow: true,
  },
})
```

**Reply by bcakmakoglu** [reply]
Or is that just your IDE making it look like that? because on my end it works with TS without any issue whatsoever 🤔

<img width="1122" alt="image" src="https://user-images.githubusercontent.com/78412429/203433238-e072e7f6-03ac-4ca9-81ca-cf7c86652585.png">

**Reply by ghiscoding** [reply]
> you can try this docs link to see the fix in action: https://vue-flow-docs-git-fix-update-model-value-bcakmakoglu.vercel.app/
> The docs always use the PR build so the fix is already available in the examples for that link

indeed that seem to do the work, you are amazingly quick in replying and fixing these issues. You are like a 🚀 😄

**Reply by ghiscoding** [reply]
> Or is that just your IDE making it look like that? because on my end it works with TS without any issue whatsoever 🤔
> 
> <img alt="image" width="1122" src="https://user-images.githubusercontent.com/78412429/203433238-e072e7f6-03ac-4ca9-81ca-cf7c86652585.png">

ahh nevermind, it was just ESLint complaining about a linting issue (indentation). my bad 😅

**Comment by bcakmakoglu** [comment, marked as answer]
The fix is released with `1.5.1` @ghiscoding

**Reply by ghiscoding** [reply]
great thanks 😄
